Initial development

shioyama opened the pull request:
This adds a basic wrapper for the OpenLogi shipping API.

I decided to use [Virtus](https://github.com/solnic/virtus) for attributes, for a few reasons:

- The OpenLogi API has explicit types associated with each object returned, which match very closely Virtus' attribute types.
- The API returns the same basic type (e.g. "item") in multiple endpoints, and also both as a single result and as collections of results. This matches well with Virtus' type system whereby you can assign an attribute the type of e.g. `Array[Openlogi::Item]` and it will handle assigning it in the background.
- Avoids some boilerplate code for assigning objects, etc.

I also opted to go with a set of "endpoints" which themselves take actions and return objects or collections of objects, like this:

```ruby
client = Openlogi::Client.new("apikey")
client.items.find(1)
```

Missing from this PR:

- [ ] Error handling
- [ ] Shipment endpoint specs
- [ ] Non-CRUD endpoint actions (image upload, bulk upload, etc.)

degicat commented:
@project-eutopia Please review this.

degicat commented:
@project-eutopia Please review this.

degicat commented:
@camelmasa Please review this.

shioyama commented:
@camelmasa I've implemented the basic actions/endpoints and added tests for items and warehousings, but not for shipments yet because I want to test the API more first.

I wasn't quite sure how far to go... so I just tried to get all the endpoints and objects and leave other stuff out for now. Actually shipments are probably most important but I'll fill out that stuff in a later pull request.

camelmasa commented:
@shioyama Sure, I'll review that :eyes:

shioyama commented:
I just added missing attributes to `Openlogi::Shipment`. There are many of them but I think it's easier just to add them all at once, rather than bit by bit.

One thing I'm hesitating about is that Virtus seems not to be supported anymore... but I quite like how easy it is to typecast and create objects from a response using it. I guess if it's an issue we could just roll out own minimal replacement within this gem. If there are alternatives out there that do a similar thing I'd be happy looking into those as well.

camelmasa commented on spec/openlogi/request_spec.rb:
```diff
@@ -0,0 +1,30 @@
+    let(:resource) { "foos" }
+    let(:response) { double("response") }
+
+    describe "GET request" do
```
@shioyama Could you add unit test for `post` pattern :question:

shioyama replied:
Ah right, forgot that. I'll do that.

shioyama replied:
I added a spec for POST requests and also included body/params.

miyazawadegica commented on lib/openlogi/shipment.rb:
```diff
@@ -0,0 +1,27 @@
+  class Shipment < BaseObject
+    attribute :id, String
+    attribute :identifier, String
+    attribute :created_at, DateTime
```
What is this? Though I saw "https://api-demo.openlogi.com/doc/api.html#出庫-出庫依頼", I could not find it. 
While, we should add `shipped_at` here. "https://api-demo.openlogi.com/doc/api.html#出庫-出庫情報"

shioyama replied:
Hmm if you look at the sample response for a shipment you will see there is a `created_at` key/value pair. I'm not actually sure what is the definitive source... but you're right that `shipped_at` is also there, so I'll add it.

miyazawadegica replied:
I missed a response. Certainly it is. `id` and `created_at` are for system value so that the document does not have them maybe.

shioyama replied:
Yes I'm including them anyway for now.

shioyama replied:
Ah `id` is not only a system value, it is used in requests. `created_at` is probably just the rails timestamp, maybe not intended to be public.

shioyama commented:
@camelmasa I filled out the remaining specs for the shipments endpoint and added some missing attributes based on the API docs. I have tried to use request/response data direct from the docs wherever possible, but in some cases (esp. shipments) the json request/responses are so big that I had to memoized in `let` blocks and modify keys depending on the request action, otherwise they would have been really huge.

There were some undocumented attributes (`created_at`) and some documented ones that did not have any examples including them, e.g. for the `delivery_options` hash I modified the response/request spec data in a `context` block to test it. It's a bit tricky because I think these things are still changing, and it's not always clear what values are expected for which request. E.g. `status` is "waiting" after a POST/PUT request to the shipments endpoint, but "shipped" for the `GET` endpoints.

Anyway I think this covers the basic functionality, could you continue reviewing it?

camelmasa commented:
@shioyama :ok_hand:  I'll continue to review.

camelmasa commented on Gemfile:
```diff
@@ -1,3 +1,6 @@
 gemspec
+
+gem "pry"
+gem "pry-byebug"
```
I think those gems should be group.

e.g.
```rb
group :development, :test do
...
end
```

or,  we should use `add_development_dependency` in `gemspec`.

shioyama replied:
Ok I wrapped them in a group. I didn't put them in the gemspec I guess because they're not really essential dependencies... you can run the specs without them, but I like to have them around. Seems like in that case people often put them in the `Gemfile` and not in the gemspec.

camelmasa replied:
I see Thanks :smiley:

camelmasa commented on openlogi.gemspec:
```diff
@@ -28,6 +28,9 @@ Gem::Specification.new do |spec|
   spec.add_dependency "typhoeus", "~> 1.0.1"
+  spec.add_dependency "activesupport"
+  spec.add_dependency "json"
+  spec.add_dependency "virtus"
```
Sorry, for tiny point.

That will be display "WARNING" when do `gem build openlogi`.
We should decide gem versions :wink:

shioyama replied:
Ah yes ok I'll add them

camelmasa commented on lib/openlogi/api/items.rb:
```diff
@@ -0,0 +1,33 @@
+        perform_request_with_objects(:get, "items", params)
+      end
+
+      def update(id, params)
```
I tried that method.

The API requires `id` and `name` params 😫

shioyama replied:
Ah you're right, I'll fix that.

shioyama replied:
Actually this doesn't really make sense... there shouldn't be a need to (re-)send the name when updating a record...

camelmasa replied:
Hmm ... little bit weird. We need to tell that to openlogi's developer :star:

shioyama replied:
Confirmed, this is not needed, they will fix it.

camelmasa replied:
Awesome :tada:

shioyama commented:
I'm wondering if we can maybe use [hashie](https://github.com/intridea/hashie) instead of virtus for coercing response hashes.

camelmasa commented:
@shioyama Could we take `Openlogi::Response` instance :question: 

e.g.
```rb
# base_object.rb
module Openlogi                                                                                                          
  class BaseObject                                                                                                       
    include Virtus.model                                                                                                 
                                                                                                                         
    attr_reader :response                                                                                                
                                                                                                                         
    def initialize(response)                                                                                             
      super(response)                                                                                                                   
      @response = response                                                                                               
    end                                                                                                                  
  end                                                                                                                    
end 
```

```rb
item = client.items.find("xxx")

if item.response.success?
  # ...
else
  # ... 
end
```

shioyama commented:
Looks like `Hashie::Dash` should work as a drop-in replacement for Virtus.

shioyama commented:
@camelmasa I switched from Virtus to [Hashie](https://github.com/intridea/hashie) with only minor changes, all existing specs pass. Hashie is a bit more flexible and closer to our needs, and also it's currently maintained whereas Virtus is not.

camelmasa commented:
Thanks, I think Hashie is nicer :smiley: I'll comment to new commits.

shioyama commented:
I added an `Enum` type in 26edc836f5a5b64ca8a6c0c0006f262c9eaeb08c, since the API docs have explicit values for many attributes. I fall through to `nil` if a value other than the expected types is returned, but I'm thinking maybe we should just fall through and return the string if it doesn't match (for whatever reason).

shioyama commented:
I updated the client so we store the response in 34b405406a238de1e0b2a3debee7c3b1cb4bf54f.

camelmasa commented:
Great :+1:  I think `Openlogi::Client#last_response` is nice idea :+1: I'll continue to test on local.

shioyama commented:
I added `error` and `errors` properties to the base object and a `valid?` method which checks whether `error` is `nil`.

Also I decided to raise an argument error if an item attribute that is an enum doesn't match any of the expected values. If that happens we should update the code to add the new value ;)

shioyama commented on lib/openlogi/boolean.rb:
```diff
@@ -0,0 +1,14 @@
+    def self.coerce(v)
+      case v
+      when String
+        !!(v =~ /^(true|t|yes|y|1)$/i)
```
@camelmasa this comes straight from the Hashie docs.

camelmasa replied:
😓 That will returns `true` when `v` is following. I think Hashie docs is wrong...

```
"foo/ntrue/nbaa"
```

shioyama replied:
Ah you mean `"foo\ntrue\nbaa"`, yes you're right, I'll fix it.

camelmasa replied:
Thanks, I'll create PR to Hashie repo :wink:

shioyama commented on lib/openlogi/api/items.rb:
```diff
@@ -0,0 +1,33 @@
+      end
+
+      def create(params)
+        perform_request_with_object(:post, "items", params)
```
I'm hesitating a bit about this, but if we wrapped `params` here with `resource_class.new(params)` we could validate parameters sent to openlogi with the same (hashie) model we use to wrap responses, so we could e.g. add required parameters, stuff like that.

shioyama replied:
Never mind, I'll leave this for another PR.

shioyama replied:
Never mind, I'll leave this for another PR.

camelmasa replied:
:wink:

camelmasa commented:
LGTM :+1: Great work :tada:  Can I merge it :question:

shioyama commented:
@camelmasa yes! thanks
